slider/beatmap.py

```diff
@@ -1816,6 +1846,127 @@ def _product_no_diagonal(sequence):
+
+        Returns
+        ----------
+        difficulties : array
+            1D array of difficulties.
+            If include_times is True, the array is 2D
+            and contains time, difficulty pairs.
+        """
+        cs = self.cs()
+        # NOTE: This is different than normal conversion
+        if hard_rock:
+            cs *= 1.3
+        elif easy:
+            cs /= 2
+        radius = circle_radius(cs)
+
+        if double_time:
+            modify = op.attrgetter('double_time')
+        elif half_time:
+            modify = op.attrgetter('half_time')
+        else:
+            def modify(e):
+                return e
+
+        if include_times:
+            strains = np.empty((len(self.hit_objects) - 1, 2))
+        else:
+            strains = np.empty(len(self.hit_objects) - 1)
+
+        hit_objects = map(modify, self.hit_objects)
+        previous = _DifficultyHitObject(next(hit_objects), radius)
+        for i, hit_object in enumerate(hit_objects):
+            new = _DifficultyHitObject(
+                hit_object,
+                radius,
+                previous,
+            )
+            if include_times:
+                strains[i] = hit_object.time.total_seconds(), new.strains[strain]
+            else:
+                strains[i] = new.strains[strain]
+            previous = new
+
+        return strains
+
+    def smoothed_difficulty(self,
+                            strain,
+                            smoothing_window,
+                            num_points,
+                            easy=False,
+                            hard_rock=False,
+                            double_time=False,
+                            half_time=False):
+        """Calculate a smoothed difficulty at evenly spaced points in time
+        between the beginning of the song and the last hit object of the map.
+
+        Done by taking an average of difficulties of hit objects within a
+        certain time window of each point
+
+        Useful if you want to calculate a difficulty curve for the map
+        since the unsmoothed values vary locally a lot.
+
+        Parameters
+        ----------
+        strain : Strain
+            Enum of the sort of strain to calculate.
+        smoothing_window : int or float
+            Time window (in seconds) for the moving average.
+            Bigger will make a smoother curve.
+        num_points : int
+            Number of points to calculate the average at.
+        easy : bool
+            Compute difficulty with easy.
+        hard_rock : bool
+            Compute difficulty with hard rock.
+        double_time : bool
+            Compute difficulty with double time.
+        half_time : bool
+            Compute difficulty with half time.
+
+        Returns
+        ----------
+        difficulties : array
+            2D array containing smoothed time, difficulty pairs.
+        """
+        return _moving_average_by_time(
```

Comment:
Is there a reason you use a simple moving average here instead of some time weighted moving average?

I think this is probably fine, I just want to open up discussion about the best way to smooth this signal.

Reply:
I just couldn't think of a better weighted version :P

If you have any suggestions I'd be happy to hear them.